File: sdk/typescript/src/namespaces/rbac.ts
/**
 * RBAC Namespace API
 *
 * Provides a namespaced interface for role-based access control operations.
 * This wraps the flat client methods for a more intuitive API.
 */

import type {
  Role,
  Permission,
  RoleAssignment,
  AssignmentList,
  BulkAssignRequest,
  BulkAssignResponse,
  PaginationParams,
} from '../types';

/**
 * Request for creating a new role.
 */
export interface CreateRoleRequest {
  /** Role name */
  name: string;
  /** Role description */
  description?: string;
  /** Permission IDs to include */
  permissions: string[];
  /** Parent role ID for inheritance */
  parent_role_id?: string;
}

/**
 * Request for updating a role.
 */
export interface UpdateRoleRequest {
  /** Updated name */
  name?: string;
  /** Updated description */
  description?: string;
  /** Updated permissions */
  permissions?: string[];
}

/**
 * Interface for the internal client methods used by RBACAPI.
 */
interface RBACClientInterface {
  // Generic request method for extended endpoints
  request<T = unknown>(method: string, path: string, options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }): Promise<T>;

  // Core RBAC methods
  listRoles(params?: PaginationParams): Promise<{ roles: Role[] }>;
  getRole(roleId: string): Promise<Role>;
  createRole(request: CreateRoleRequest): Promise<Role>;
  updateRole(roleId: string, updates: UpdateRoleRequest): Promise<Role>;
  deleteRole(roleId: string): Promise<{ deleted: boolean }>;
  listPermissions(params?: PaginationParams): Promise<{ permissions: Permission[] }>;
  assignRole(userId: string, roleId: string): Promise<{ assigned: boolean }>;
  revokeRole(userId: string, roleId: string): Promise<void>;
  getUserRoles(userId: string): Promise<{ roles: Role[] }>;
  checkPermission(userId: string, permission: string): Promise<{ allowed: boolean }>;
  listRoleAssignments(roleId: string, params?: PaginationParams): Promise<AssignmentList>;
  bulkAssignRoles(body: BulkAssignRequest): Promise<BulkAssignResponse>;
}

/**
 * RBAC API namespace.
 *
 * Provides methods for role-based access control:
 * - Managing roles and permissions
 * - Assigning and revoking roles
 * - Checking user permissions
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // List all roles
 * const { roles } = await client.rbac.listRoles();
 *
 * // Create a new role
 * const role = await client.rbac.createRole({
 *   name: 'Analyst',
 *   permissions: ['debates:read', 'analytics:read'],
 * });
 *
 * // Assign role to user
 * await client.rbac.assignRole(userId, role.id);
 *
 * // Check permission
 * const { allowed } = await client.rbac.checkPermission(userId, 'debates:create');
 * ```
 */
export class RBACAPI {
  constructor(private client: RBACClientInterface) {}

  /**
   * List all roles with optional pagination.
   */
  async listRoles(params?: PaginationParams): Promise<{ roles: Role[] }> {
    return this.client.listRoles(params);
  }

  /**
   * Get a role by ID.
   */
  async getRole(roleId: string): Promise<Role> {
    return this.client.getRole(roleId);
  }

  /**
   * Create a new role.
   */
  async createRole(request: CreateRoleRequest): Promise<Role> {
    return this.client.createRole(request);
  }

  /**
   * Update an existing role.
   */
  async updateRole(roleId: string, updates: UpdateRoleRequest): Promise<Role> {
    return this.client.updateRole(roleId, updates);
  }

  /**
   * Delete a role.
   */
  async deleteRole(roleId: string): Promise<{ deleted: boolean }> {
    return this.client.deleteRole(roleId);
  }

  /**
   * List all permissions with optional pagination.
   */
  async listPermissions(params?: PaginationParams): Promise<{ permissions: Permission[] }> {
    return this.client.listPermissions(params);
  }

  /**
   * Assign a role to a user.
   */
  async assignRole(userId: string, roleId: string): Promise<{ assigned: boolean }> {
    await this.client.assignRole(userId, roleId);
    return { assigned: true };
  }

  /**
   * Revoke a role from a user.
   */
  async revokeRole(userId: string, roleId: string): Promise<void> {
    return this.client.revokeRole(userId, roleId);
  }

  /**
   * Get all roles assigned to a user.
   */
  async getUserRoles(userId: string): Promise<{ roles: Role[] }> {
    return this.client.getUserRoles(userId);
  }

  /**
   * Check if a user has a specific permission.
   */
  async checkPermission(userId: string, permission: string): Promise<{ allowed: boolean }> {
    return this.client.checkPermission(userId, permission);
  }

  /**
   * List users assigned to a role.
   */
  async listAssignments(
    roleId: string,
    params?: PaginationParams
  ): Promise<AssignmentList> {
    return this.client.listRoleAssignments(roleId, params);
  }

  /**
   * Bulk assign roles to multiple users.
   */
  async bulkAssign(body: BulkAssignRequest): Promise<BulkAssignResponse> {
    return this.client.bulkAssignRoles(body);
  }

  // =========================================================================
  // User Management
  // =========================================================================

  /**
   * List users in organization.
   */
  async listUsers(params?: PaginationParams): Promise<{ users: unknown[]; total: number }> {
    return this.client.request('GET', '/api/users', { params });
  }

  /**
   * Invite a new user to the organization.
   */
  async inviteUser(email: string, role?: string): Promise<{ invitation_id: string; email: string }> {
    return this.client.request('POST', '/api/users/invite', { json: { email, role } });
  }

  /**
   * Remove a user from the organization.
   */
  async removeUser(userId: string): Promise<{ removed: boolean }> {
    return this.client.request('DELETE', `/api/users/${userId}`);
  }

  /**
   * Change a user's role in the organization.
   */
  async changeUserRole(userId: string, role: string): Promise<{ updated: boolean }> {
    return this.client.request('PUT', `/api/users/${userId}/role`, { json: { role } });
  }

  // =========================================================================
  // Workspace Roles
  // =========================================================================

  /**
   * Get available roles for a workspace based on RBAC profile.
   */
  async getWorkspaceRoles(workspaceId: string): Promise<{ roles: unknown[]; profile: string }> {
    return this.client.request('GET', `/api/v1/workspaces/${workspaceId}/roles`);
  }

  /**
   * Update a member's role in a workspace.
   */
  async updateMemberRole(workspaceId: string, userId: string, role: string): Promise<{ updated: boolean }> {
    return this.client.request('PUT', `/api/v1/workspaces/${workspaceId}/members/${userId}/role`, { json: { role } });
  }

  /**
   * Add a member to a workspace.
   */
  async addWorkspaceMember(workspaceId: string, userId: string, role?: string): Promise<{ added: boolean }> {
    return this.client.request('POST', `/api/v1/workspaces/${workspaceId}/members`, { json: { user_id: userId, role } });
  }

  /**
   * Remove a member from a workspace.
   */
  async removeWorkspaceMember(workspaceId: string, userId: string): Promise<{ removed: boolean }> {
    return this.client.request('DELETE', `/api/v1/workspaces/${workspaceId}/members/${userId}`);
  }

  /**
   * List available RBAC profiles.
   */
  async listProfiles(): Promise<{ profiles: unknown[] }> {
    return this.client.request('GET', '/api/v1/workspaces/profiles');
  }

  // =========================================================================
  // Audit
  // =========================================================================

  /**
   * Query privacy audit entries.
   */
  async queryAudit(options?: { action?: string; user_id?: string; since?: string; limit?: number }): Promise<{ entries: unknown[]; total: number }> {
    return this.client.request('GET', '/api/v1/audit/entries', { params: options });
  }

  /**
   * Generate compliance audit report.
   */
  async getAuditReport(options?: { framework?: string; since?: string }): Promise<unknown> {
    return this.client.request('GET', '/api/v1/audit/report', { params: options });
  }

  /**
   * Verify audit log integrity.
   */
  async verifyAuditIntegrity(): Promise<{ valid: boolean; issues: unknown[] }> {
    return this.client.request('GET', '/api/v1/audit/verify');
  }

  /**
   * Get user activity history.
   */
  async getUserActivityHistory(userId: string, options?: PaginationParams): Promise<{ activities: unknown[]; total: number }> {
    return this.client.request('GET', `/api/v1/audit/actor/${userId}/history`, { params: options });
  }

  /**
   * Get resource access history.
   */
  async getResourceHistory(resourceType: string, resourceId: string, options?: PaginationParams): Promise<{ accesses: unknown[]; total: number }> {
    return this.client.request('GET', `/api/v1/audit/resource/${resourceType}/${resourceId}/history`, { params: options });
  }

  /**
   * Get denied access attempts.
   */
  async getDeniedAccess(options?: PaginationParams): Promise<{ denied: unknown[]; total: number }> {
    return this.client.request('GET', '/api/v1/audit/denied', { params: options });
  }

  // =========================================================================
  // API Keys
  // =========================================================================

  /**
   * Generate a new API key.
   */
  async generateApiKey(name: string, permissions?: string[], expires_at?: string): Promise<{ key: string; key_id: string }> {
    return this.client.request('POST', '/api/auth/api-key', { json: { name, permissions, expires_at } });
  }

  /**
   * List API keys.
   */
  async listApiKeys(): Promise<{ keys: unknown[] }> {
    return this.client.request('GET', '/api/keys');
  }

  /**
   * Revoke an API key.
   */
  async revokeApiKey(keyId: string): Promise<{ revoked: boolean }> {
    return this.client.request('DELETE', `/api/keys/${keyId}`);
  }

  // =========================================================================
  // Sessions
  // =========================================================================

  /**
   * List active sessions for the current user.
   */
  async listSessions(): Promise<{ sessions: unknown[] }> {
    return this.client.request('GET', '/api/auth/sessions');
  }

  /**
   * Revoke a specific session.
   */
  async revokeSession(sessionId: string): Promise<{ revoked: boolean }> {
    return this.client.request('DELETE', `/api/auth/sessions/${sessionId}`);
  }

  /**
   * Logout from all devices.
   */
  async logoutAll(): Promise<{ logged_out: boolean }> {
    return this.client.request('POST', '/api/auth/logout-all');
  }

  // =========================================================================
  // MFA
  // =========================================================================

  /**
   * Setup MFA - generate secret and QR code.
   */
  async setupMfa(): Promise<{ secret: string; qr_code: string }> {
    return this.client.request('POST', '/api/auth/mfa/setup');
  }

  /**
   * Enable MFA by verifying setup code.
   */
  async enableMfa(code: string): Promise<{ enabled: boolean; backup_codes: string[] }> {
    return this.client.request('POST', '/api/auth/mfa/enable', { json: { code } });
  }

  /**
   * Disable MFA.
   */
  async disableMfa(code: string): Promise<{ disabled: boolean }> {
    return this.client.request('POST', '/api/auth/mfa/disable', { json: { code } });
  }

  /**
   * Verify MFA code during login.
   */
  async verifyMfa(code: string): Promise<{ verified: boolean; token?: string }> {
    return this.client.request('POST', '/api/auth/mfa/verify', { json: { code } });
  }

  /**
   * Regenerate MFA backup codes.
   */
  async regenerateBackupCodes(code: string): Promise<{ backup_codes: string[] }> {
    return this.client.request('POST', '/api/auth/mfa/backup-codes', { json: { code } });
  }
}

File: sdk/python/aragora_sdk/namespaces/rbac.py
"""
RBAC Namespace API

Provides methods for Role-Based Access Control:
- Permission management
- Role management
- Role assignments
- Permission checking
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient

class RBACAPI:
    """
    Synchronous RBAC API.

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> roles = client.rbac.list_roles()
        >>> can_access = client.rbac.check_permission("user_123", "debates:create")
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    def list_permissions(
        self,
        resource_type: str | None = None,
        limit: int = 100,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List available permissions.

        Args:
            resource_type: Filter by resource type
            limit: Maximum results
            offset: Pagination offset

        Returns:
            List of permissions with descriptions
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if resource_type:
            params["resource_type"] = resource_type

        return self._client.request("GET", "/api/v1/rbac/permissions", params=params)

    def get_permission(self, permission_key: str) -> dict[str, Any]:
        """
        Get a permission by key.

        Args:
            permission_key: Permission key (e.g., "debates:create")

        Returns:
            Permission details
        """
        return self._client.request("GET", f"/api/v1/rbac/permissions/{permission_key}")

    def list_roles(
        self,
        include_system: bool = True,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List available roles.

        Args:
            include_system: Include system roles (admin, viewer, etc.)
            limit: Maximum results
            offset: Pagination offset

        Returns:
            List of roles with permissions
        """
        params: dict[str, Any] = {
            "include_system": include_system,
            "limit": limit,
            "offset": offset,
        }
        return self._client.request("GET", "/api/v1/rbac/roles", params=params)

    def get_role(self, role_id: str) -> dict[str, Any]:
        """
        Get a role by ID.

        Args:
            role_id: Role ID

        Returns:
            Role details with permissions
        """
        return self._client.request("GET", f"/api/v1/rbac/roles/{role_id}")

    def create_role(
        self,
        name: str,
        permissions: list[str],
        description: str | None = None,
        parent_role: str | None = None,
    ) -> dict[str, Any]:
        """
        Create a custom role.

        Args:
            name: Role name
            permissions: List of permission keys
            description: Role description
            parent_role: Parent role to inherit from

        Returns:
            Created role
        """
        data: dict[str, Any] = {"name": name, "permissions": permissions}
        if description:
            data["description"] = description
        if parent_role:
            data["parent_role"] = parent_role

        return self._client.request("POST", "/api/v1/rbac/roles", json=data)

    def update_role(
        self,
        role_id: str,
        permissions: list[str] | None = None,
        description: str | None = None,
    ) -> dict[str, Any]:
        """
        Update a role.

        Args:
            role_id: Role ID
            permissions: New permissions list
            description: New description

        Returns:
            Updated role
        """
        data: dict[str, Any] = {}
        if permissions is not None:
            data["permissions"] = permissions
        if description is not None:
            data["description"] = description

        return self._client.request("PUT", f"/api/v1/rbac/roles/{role_id}", json=data)

    def delete_role(self, role_id: str) -> dict[str, Any]:
        """
        Delete a custom role.

        Args:
            role_id: Role ID

        Returns:
            Deletion result
        """
        return self._client.request("DELETE", f"/api/v1/rbac/roles/{role_id}")

    def assign_role(
        self,
        user_id: str,
        role_id: str,
        scope: str | None = None,
    ) -> dict[str, Any]:
        """
        Assign a role to a user.

        Args:
            user_id: User ID
            role_id: Role ID to assign
            scope: Optional scope (org, workspace, etc.)

        Returns:
            Role assignment
        """
        data: dict[str, Any] = {"user_id": user_id, "role_id": role_id}
        if scope:
            data["scope"] = scope

        return self._client.request("POST", "/api/v1/rbac/assignments", json=data)

    def revoke_role(self, user_id: str, role_id: str) -> dict[str, Any]:
        """
        Revoke a role from a user.

        Args:
            user_id: User ID
            role_id: Role ID to revoke

        Returns:
            Revocation result
        """
        return self._client.request(
            "DELETE",
            "/api/v1/rbac/assignments",
            params={"user_id": user_id, "role_id": role_id},
        )

    def check_permission(
        self,
        user_id: str,
        permission: str,
        resource_id: str | None = None,
    ) -> dict[str, Any]:
        """
        Check if a user has a specific permission.

        Args:
            user_id: User ID
            permission: Permission key to check
            resource_id: Optional specific resource

        Returns:
            Permission check result with allowed status
        """
        data: dict[str, Any] = {"user_id": user_id, "permission": permission}
        if resource_id:
            data["resource_id"] = resource_id

        return self._client.request("POST", "/api/v1/rbac/check", json=data)

    def list_profiles(self) -> dict[str, Any]:
        """List available RBAC profiles (lite, standard, enterprise)."""
        return self._client.request("GET", "/api/v1/workspaces/profiles")

    # ========== Audit Trail ==========

    def query_audit(
        self,
        action: str | None = None,
        user_id: str | None = None,
        since: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """Query audit log entries."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if action:
            params["action"] = action
        if user_id:
            params["user_id"] = user_id
        if since:
            params["since"] = since
        return self._client.request("GET", "/api/v1/audit/entries", params=params)

    def get_audit_report(
        self, framework: str | None = None, since: str | None = None
    ) -> dict[str, Any]:
        """Generate compliance audit report."""
        params: dict[str, Any] = {}
        if framework:
            params["framework"] = framework
        if since:
            params["since"] = since
        return self._client.request("GET", "/api/v1/audit/report", params=params)

    def verify_audit_integrity(self) -> dict[str, Any]:
        """Verify audit log integrity."""
        return self._client.request("GET", "/api/v1/audit/verify")

    def get_denied_access(self, limit: int = 50, offset: int = 0) -> dict[str, Any]:
        """Get denied access attempts."""
        return self._client.request(
            "GET", "/api/v1/audit/denied", params={"limit": limit, "offset": offset}
        )

    # ========== API Keys ==========

    def generate_api_key(
        self,
        name: str,
        permissions: list[str] | None = None,
        expires_at: str | None = None,
    ) -> dict[str, Any]:
        """Generate a new API key."""
        payload: dict[str, Any] = {"name": name}
        if permissions:
            payload["permissions"] = permissions
        if expires_at:
            payload["expires_at"] = expires_at
        return self._client.request("POST", "/api/auth/api-key", json=payload)

    def list_api_keys(self) -> dict[str, Any]:
        """List API keys for current user."""
        return self._client.request("GET", "/api/keys")

    def revoke_api_key(self, key_id: str) -> dict[str, Any]:
        """Revoke an API key."""
        return self._client.request("DELETE", f"/api/keys/{key_id}")

    # ========== Sessions ==========

    def list_sessions(self) -> dict[str, Any]:
        """List active sessions for current user."""
        return self._client.request("GET", "/api/auth/sessions")

    def revoke_session(self, session_id: str) -> dict[str, Any]:
        """Revoke a specific session."""
        return self._client.request("DELETE", f"/api/auth/sessions/{session_id}")

    def logout_all(self) -> dict[str, Any]:
        """Logout from all devices."""
        return self._client.request("POST", "/api/auth/logout-all", json={})

    # ========== MFA ==========

    def setup_mfa(self) -> dict[str, Any]:
        """Setup MFA - generate secret and QR code."""
        return self._client.request("POST", "/api/auth/mfa/setup", json={})

    def enable_mfa(self, code: str) -> dict[str, Any]:
        """Enable MFA by verifying setup code."""
        return self._client.request("POST", "/api/auth/mfa/enable", json={"code": code})

    def disable_mfa(self, code: str) -> dict[str, Any]:
        """Disable MFA."""
        return self._client.request("POST", "/api/auth/mfa/disable", json={"code": code})

    def verify_mfa(self, code: str) -> dict[str, Any]:
        """Verify MFA code during login."""
        return self._client.request("POST", "/api/auth/mfa/verify", json={"code": code})

    def regenerate_backup_codes(self, code: str) -> dict[str, Any]:
        """Regenerate MFA backup codes."""
        return self._client.request("POST", "/api/auth/mfa/backup-codes", json={"code": code})

class AsyncRBACAPI:
    """
    Asynchronous RBAC API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     roles = await client.rbac.list_roles()
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def list_permissions(
        self,
        resource_type: str | None = None,
        limit: int = 100,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List available permissions."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if resource_type:
            params["resource_type"] = resource_type

        return await self._client.request("GET", "/api/v1/rbac/permissions", params=params)

    async def get_permission(self, permission_key: str) -> dict[str, Any]:
        """Get a permission by key."""
        return await self._client.request("GET", f"/api/v1/rbac/permissions/{permission_key}")

    async def list_roles(
        self,
        include_system: bool = True,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List available roles."""
        params: dict[str, Any] = {
            "include_system": include_system,
            "limit": limit,
            "offset": offset,
        }
        return await self._client.request("GET", "/api/v1/rbac/roles", params=params)

    async def get_role(self, role_id: str) -> dict[str, Any]:
        """Get a role by ID."""
        return await self._client.request("GET", f"/api/v1/rbac/roles/{role_id}")

    async def create_role(
        self,
        name: str,
        permissions: list[str],
        description: str | None = None,
        parent_role: str | None = None,
    ) -> dict[str, Any]:
        """Create a custom role."""
        data: dict[str, Any] = {"name": name, "permissions": permissions}
        if description:
            data["description"] = description
        if parent_role:
            data["parent_role"] = parent_role

        return await self._client.request("POST", "/api/v1/rbac/roles", json=data)

    async def update_role(
        self,
        role_id: str,
        permissions: list[str] | None = None,
        description: str | None = None,
    ) -> dict[str, Any]:
        """Update a role."""
        data: dict[str, Any] = {}
        if permissions is not None:
            data["permissions"] = permissions
        if description is not None:
            data["description"] = description

        return await self._client.request("PUT", f"/api/v1/rbac/roles/{role_id}", json=data)

    async def delete_role(self, role_id: str) -> dict[str, Any]:
        """Delete a custom role."""
        return await self._client.request("DELETE", f"/api/v1/rbac/roles/{role_id}")

    async def assign_role(
        self,
        user_id: str,
        role_id: str,
        scope: str | None = None,
    ) -> dict[str, Any]:
        """Assign a role to a user."""
        data: dict[str, Any] = {"user_id": user_id, "role_id": role_id}
        if scope:
            data["scope"] = scope

        return await self._client.request("POST", "/api/v1/rbac/assignments", json=data)

    async def revoke_role(self, user_id: str, role_id: str) -> dict[str, Any]:
        """Revoke a role from a user."""
        return await self._client.request(
            "DELETE",
            "/api/v1/rbac/assignments",
            params={"user_id": user_id, "role_id": role_id},
        )

    async def check_permission(
        self,
        user_id: str,
        permission: str,
        resource_id: str | None = None,
    ) -> dict[str, Any]:
        """Check if a user has a specific permission."""
        data: dict[str, Any] = {"user_id": user_id, "permission": permission}
        if resource_id:
            data["resource_id"] = resource_id

        return await self._client.request("POST", "/api/v1/rbac/check", json=data)

    async def list_profiles(self) -> dict[str, Any]:
        """List available RBAC profiles (lite, standard, enterprise)."""
        return await self._client.request("GET", "/api/v1/workspaces/profiles")

    # ========== Audit Trail ==========

    async def query_audit(
        self,
        action: str | None = None,
        user_id: str | None = None,
        since: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """Query audit log entries."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if action:
            params["action"] = action
        if user_id:
            params["user_id"] = user_id
        if since:
            params["since"] = since
        return await self._client.request("GET", "/api/v1/audit/entries", params=params)

    async def get_audit_report(
        self, framework: str | None = None, since: str | None = None
    ) -> dict[str, Any]:
        """Generate compliance audit report."""
        params: dict[str, Any] = {}
        if framework:
            params["framework"] = framework
        if since:
            params["since"] = since
        return await self._client.request("GET", "/api/v1/audit/report", params=params)

    async def verify_audit_integrity(self) -> dict[str, Any]:
        """Verify audit log integrity."""
        return await self._client.request("GET", "/api/v1/audit/verify")

    async def get_denied_access(self, limit: int = 50, offset: int = 0) -> dict[str, Any]:
        """Get denied access attempts."""
        return await self._client.request(
            "GET", "/api/v1/audit/denied", params={"limit": limit, "offset": offset}
        )

    # ========== API Keys ==========

    async def generate_api_key(
        self,
        name: str,
        permissions: list[str] | None = None,
        expires_at: str | None = None,
    ) -> dict[str, Any]:
        """Generate a new API key."""
        payload: dict[str, Any] = {"name": name}
        if permissions:
            payload["permissions"] = permissions
        if expires_at:
            payload["expires_at"] = expires_at
        return await self._client.request("POST", "/api/auth/api-key", json=payload)

    async def list_api_keys(self) -> dict[str, Any]:
        """List API keys for current user."""
        return await self._client.request("GET", "/api/keys")

    async def revoke_api_key(self, key_id: str) -> dict[str, Any]:
        """Revoke an API key."""
        return await self._client.request("DELETE", f"/api/keys/{key_id}")

    # ========== Sessions ==========

    async def list_sessions(self) -> dict[str, Any]:
        """List active sessions for current user."""
        return await self._client.request("GET", "/api/auth/sessions")

    async def revoke_session(self, session_id: str) -> dict[str, Any]:
        """Revoke a specific session."""
        return await self._client.request("DELETE", f"/api/auth/sessions/{session_id}")

    async def logout_all(self) -> dict[str, Any]:
        """Logout from all devices."""
        return await self._client.request("POST", "/api/auth/logout-all", json={})

    # ========== MFA ==========

    async def setup_mfa(self) -> dict[str, Any]:
        """Setup MFA - generate secret and QR code."""
        return await self._client.request("POST", "/api/auth/mfa/setup", json={})

    async def enable_mfa(self, code: str) -> dict[str, Any]:
        """Enable MFA by verifying setup code."""
        return await self._client.request("POST", "/api/auth/mfa/enable", json={"code": code})

    async def disable_mfa(self, code: str) -> dict[str, Any]:
        """Disable MFA."""
        return await self._client.request("POST", "/api/auth/mfa/disable", json={"code": code})

    async def verify_mfa(self, code: str) -> dict[str, Any]:
        """Verify MFA code during login."""
        return await self._client.request("POST", "/api/auth/mfa/verify", json={"code": code})

    async def regenerate_backup_codes(self, code: str) -> dict[str, Any]:
        """Regenerate MFA backup codes."""
        return await self._client.request("POST", "/api/auth/mfa/backup-codes", json={"code": code})
